### PR TITLE
fix: emit GFM-safe Markdown

### LINKED ISSUE #146: Block-level content inside `<blockquote>` collapses onto one line or escapes the quote
## Problem 

Spacing is zeroed for every element inside a `<blockquote>`, so block-level children lose their line breaks and `>` continuation. They run together on one line, and multi-line blocks drop the `>` after their first line, which ends the quote mid-block. 

**Reproduction**

```html
<blockquote><ul><li>one</li><li>two</li></ul></blockquote>
```

Actual:

```
> - one - two
```

Expected:

```
> - one
> - two
```

Affected blocks, same cause. Each case below shows the input, then the current output, then the fixed output:

**`<blockquote><ol><li>one</li><li>two</li></ol></blockquote>`**

```
> 1. one 2. two
```
```
> 1. one
> 2. two
```

**`<blockquote><p>intro</p><ul><li>one</li></ul></blockquote>`**

```
> intro - one
```
```
> intro
>
> - one
```

**`<blockquote>text<h2>H</h2></blockquote>`** — Actual parses as a paragraph, not a heading:

```
> text ## H
```
```
> text
>
> ## H
```

**`<blockquote>a<hr>b</blockquote>`** — Actual collapses to one paragraph, losing the rule:

```
> a --- b
```
```
> a
>
> ---
> b
```

**`<blockquote><div>a</div><div>b</div></blockquote>`** — Actual merges both divs into one paragraph:

```
> a b
```
```
> a
>
> b
```

**`<blockquote>lead<table>…</table>tail</blockquote>`** — Actual: the table row leaks out of the quote and `tail` is eaten as the delimiter row:

```
> lead | a |
| --- | tail
```
```
> lead
>
> | a |
> | --- |
>
> tail
```

**`<blockquote>lead<section>x</section>tail</blockquote>`** — same for <article>, <nav>, and <figure>, all glued to preceding text the same way

```
> lead x tail
```
```
> lead
>
> x
> tail
```

**`<blockquote><ul><li>one<ul><li>sub</li></ul></li></ul></blockquote>`** — nested list indent is destroyed:

```
> - one  - sub
```
```
> - one
>   - sub
```

**`<ul><li><blockquote><ul>…</ul></blockquote></li></ul>`** — blockquote nested inside a list item:

```
- 
  >   - x  - y
```
```
- 
  > - x
  > - y
```

### LINKED ISSUE #149: Backticks inside `<code>` and `<pre>` are backslash-escaped
## Problem

In GFM, backslash escapes are inert inside code spans and fenced code blocks: the `\` is emitted literally. mdream backslash-escapes every backtick in code context, so any code that legitimately contains backticks (shell snippets, docs about Markdown, nested fences) renders with stray backslashes.

## Reproduction 

```html
<p>Inline with backticks: <code>a `b` c</code>.</p>
<pre><code>Contains ```triple backticks``` inside.</code></pre>
```
Actual: `` `a \`b\` c` ``;  `` Contains \`\`\`triple backticks\`\`\` inside. ``.

Expected:

```
`` a `b` c ``
```

````
```
Contains ```triple backticks``` inside.
```
````

### LINKED ISSUE #164: `<br>` in prose emits a soft break, lost as `<br>` under default GFM rendering
Following #128 / #130 / #140. `<br>` in prose currently emits a bare `\n`:

```
htmlToMarkdown("<p>line one<br>line two</p>")  →  "line one\nline two"
```

Per GFM §6.13, a line break not preceded by two spaces or a backslash is a *soft* break, and "a softbreak may be rendered in HTML either as a line ending or as a space. The result will be the same in browsers." So under default GFM rendering the `<br>` collapses to a space (verified with `cmark-gfm`, the GFM reference implementation):

```
$ printf 'line one\nline two\n' | cmark-gfm
<p>line one
line two</p>          # displays as "line one line two" — the break is gone
```

GFM §6.12 says only two trailing spaces or a trailing backslash produce a hard break (`<br />`). Both keep a newline in the source, so the raw text still reads on two lines:

```
$ printf 'line one  \nline two\n' | cmark-gfm     # two trailing spaces
<p>line one<br />
line two</p>

$ printf 'line one\\\nline two\n' | cmark-gfm      # backslash
<p>line one<br />
line two</p>
```

Caveat: GitHub's own issue/comment renderer enables the optional "soft breaks as hard breaks" setting from §6.13, so a bare `\n` *does* show as a break in a GitHub comment. That's a non-default renderer option, though. README/`.md` rendering on GitHub and other GFM consumers follow the spec default and drop the break.

### PR DESCRIPTION
**Pull request by harlan-zw**
### 🔗 Linked issue

Resolves #146
Resolves #149
Resolves #164

### ❓ Type of change

- [ ] 📖 Documentation
- [x] 🐞 Bug fix
- [ ] 👌 Enhancement
- [ ] ✨ New feature
- [ ] 🧹 Chore
- [ ] ⚠️ Breaking change

### 📚 Description

Literal text, links, code fences, line breaks, and nested blockquotes could reparse into different GFM structures. The JavaScript and Rust engines now serialize those cases safely, with one-shot and streaming parity coverage.

The audit parsed all 648 official GFM examples, converted their expected HTML through both engines, then reparsed the generated Markdown. Focused regressions cover the five confirmed serializer problem groups.

### 📊 Performance

The hot path now records one Rust provenance bit while parsing and uses a native JavaScript prefilter before the precise escaper. Append-only Rust resource writers avoid temporary serializer strings. On the 1.8 MB wiki fixture, only 633 of 31,111 JavaScript text nodes enter the precise scanner.

Local paired runs against current main measure +5.5% JavaScript core CPU, +3.7% stream CPU, and +7.7% Rust WASM CPU. Semantic allocations are +3.3% core and +2.8% stream, with unchanged WASM peak memory. This is down from the first CI revision at +16.6%, +17.5%, and +20.9%. The bundle analysis check will publish the authoritative shared-run comparison.

### ✅ Verification

- `pnpm test`
- `cargo test -p mdream --manifest-path crates/Cargo.toml`
- `pnpm lint`
- `pnpm typecheck`
- `pnpm build`

## Summary by CodeRabbit

* **New Features**
  * Enhanced GFM-aware escaping and safer decoded-entity handling to preserve reparsable Markdown.
  * Improved link and image Markdown serialization (better escaping, whitespace URLs via angle-bracket destinations, literal alt/title handling).
  * More robust code rendering with automatic delimiter widening, plus refined blockquote and hard-break (`<br>`) output behavior.

* **Bug Fixes**
  * Fixed Markdown streaming correctness across chunk boundaries for code fences/spans and deeply nested blockquotes.
  * Prevented double-escaping and malformed Markdown in tables, entities, and escaping edge cases.
  * Clean mode now strips executable/meaningless link schemes case-insensitively.

* **Tests**
  * Expanded unit and streaming-parity regression coverage for escaping, links/images, entities, code delimiters, blockquotes, and hard breaks.

### CODE REVIEW

File: packages/js/src/tags.ts [hunk 1]
```diff
@@ -157,6 +157,54 @@ export function resolveUrl(url: string, origin?: string): string {
...
+
+function stripsEmptyLink(state: HandlerContext['state'], href: string): boolean {
+  const clean = state.options?.clean
+  if (!(clean === true || (typeof clean === 'object' && clean.emptyLinks)))
+    return false
+  return href === '#'
+    || href.startsWith('javascript:')
+    || href.startsWith('data:')
+    || href.startsWith('vbscript:')
+}
```
[the remark this answers is not in the document]
Comment (harlan-zw) [reply packages/js/src/tags.ts:175]: Fixed in 430437c. Rust clean.emptyLinks now strips data: and vbscript: links to match JavaScript, with a regression test.

File: packages/js/test/unit/gfm-text-escaping.test.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+): Promise<string> {
+  const stream = new ReadableStream<string>({
+    start(controller) {
+      for (let index = 0; index < html.length; index += chunkSize)
+        controller.enqueue(html.slice(index, index + chunkSize))
+      controller.close()
+    },
+  })
+  let output = ''
+  for await (const chunk of streamHtmlToMarkdown(stream, options))
```
[the remark this answers is not in the document]
Comment (harlan-zw) [reply packages/js/test/unit/gfm-text-escaping.test.ts:18]: Verified as a false positive. This test imports the public wrapper from src/index, which resolves options.hooks and passes the resolved plugins to the low-level stream converter. The existing streamed replaceText assertion passes only when the plugin executes; targeted coverage is 21/21 green.

File: crates/core/src/convert/output.rs [hunk 3]
```diff
@@ -134,18 +457,29 @@ impl ConvertState {
...
-        // emptyLinks: skip href="#" or "javascript:"
+        // emptyLinks: skip hrefs that cannot represent meaningful navigation.
         if self.clean_flags & CLEAN_EMPTY_LINKS != 0 {
           let node = &self.stack[self.stack.len() - 1];
           if let Some(href) = node.attributes.get("href")
-            && (href == "#" || href.starts_with("javascript:"))
+            && (href == "#"
+              || href.starts_with("javascript:")
+              || href.starts_with("data:")
+              || href.starts_with("vbscript:"))
```
[the remark this answers is not in the document]
Comment (harlan-zw) [reply crates/core/src/convert/output.rs]: Fixed in be01395. Rust and JavaScript now detect JavaScript:, DATA:, and VbScRiPt: case-insensitively without lowercase allocations. Added failing-first coverage for both serializers and the JavaScript postprocessor.

**Comment by harlan-zw** [comment]
Review follow-up in be01395: spacing-only blockquote overrides now retain the default quote frame. Literal enter and exit overrides still replace it; both cases have regression coverage.
